BFAIR/normalization/normalization.py

```diff
@@ -0,0 +1,297 @@
+import pandas as pd
+import numpy as np
+import copy
+
+
+def min_max_norm(
+    df, columnname="Intensity", groupname_colname="sample_group_name"
+):
+    """
+    Function that applies min/max scaling to the input dataframe; the data
+    will range from 0 for the min- to 1 for the max value
+
+    Parameters
+    ----------
+    df: pandas.DataFrame
+        input dataframe, output of either the extractNamesAndIntensities()
+        or the calculateMeanVarRSD() function
+    columnname: string
+        the name of the column with the data that needs to be
+        normalized, defaults to 'Intensity'
+    groupname_colname: string
+        the name of the column with the sample group names,
+        defaults to 'sample_group_name'
+
+    Returns
+    -------
+    output_df: pandas.DataFrame
+        the output dataframe. It follows the same architecture as
+        the input dataframe, just with normalized values
+    """
+    output_df = pd.DataFrame()
+    sample_group_names = df[groupname_colname].unique()
+    for i, sample_group_name in enumerate(sample_group_names):
+        new_df = copy.deepcopy(df[df[groupname_colname] == sample_group_name])
+        min_val = min(new_df[columnname])
+        max_val = max(new_df[columnname])
+        new_df[columnname] = new_df[columnname].sub(min_val)
+        new_df[columnname] = new_df[columnname].div((max_val - min_val))
+        if i == 0:
+            output_df = new_df
+        else:
+            output_df = output_df.append(new_df)
+    return output_df
+
+
+def tsi_norm(
+    df, columnname="Intensity", groupname_colname="sample_group_name"
+):
+    """
+    Applies Total Sum Intensity normalization; all values will be divided by
+    the sum of all values. Like that, summing up all the values would
+    amount to '1'
+
+    Parameters
+    ----------
+    df: pandas.DataFrame
+        input dataframe, output of either the extractNamesAndIntensities()
+        or the calculateMeanVarRSD() function
+    columnname: string
+        the name of the column with the data that needs to be
+        normalized, defaults to 'Intensity'
+    groupname_colname: string
+        the name of the column with the sample group names,
+        defaults to 'sample_group_name'
+
+    Returns
+    -------
+    output_df: pandas.DataFrame
+        the output dataframe. It follows the same architecture as
+        the input dataframe, just with normalized values
+    """
+    output_df = pd.DataFrame()
+    sample_group_names = df[groupname_colname].unique()
+    for i, sample_group_name in enumerate(sample_group_names):
+        new_df = copy.deepcopy(df[df[groupname_colname] == sample_group_name])
+        tsi = new_df[columnname].sum(skipna=True)
+        new_df[columnname] = new_df[columnname].div(tsi)
+        if i == 0:
+            output_df = new_df
+        else:
+            output_df = output_df.append(new_df)
+    return output_df
+
+
+def lim_tsi_norm(
```

Comment:
It looks like we are getting closer.  However, I am still not sure why the parameters for `lim_type`, `product_df`, and `biomass_value` are needed.  Essentially, the method is just normalizing to a select set of metabolites with or without weights.  It seems that this could be achieved by only using the `metabolite_input` dataframe and providing the appropriate weightings.

Reply:
So I kept the products separate from the substrates because I didn't include them in the unweighted normalization; I guess I could add them with a negative coefficient to make things easier. The biomass_value is the coefficient of the product biomass in the biomass function. I'd prefer to keep it separate rather then adding it to the df just to not include a bunch of if statements in other steps and to prevent confusion. I added the "lim_type" because I thought that it's easier to just let the user pick. If you think that I should make it dependent on the input (i.e. if it's a list then it will only pick the metabolites and normalize according to them, if it's a df then it will apply weights as well) then I can do that. Like that the amino acid and biomass versions would perform the same operation